mongo/integration/clam_prose_test.go

```diff
@@ -0,0 +1,246 @@
+					5+len(logger.TruncationSuffix), len(cmd))
+			}
+
+			return nil
+		}),
+		newLogTruncCaseValidator(mt, "reply", func(cmd string) error {
+			if len(cmd) != 5+len(logger.TruncationSuffix) {
+				return fmt.Errorf("expected reply to be %d bytes, got %d",
+					5+len(logger.TruncationSuffix), len(cmd))
+			}
+
+			return nil
+		}),
+	}
+}
+
+func clamExplicitTruncLimitFailOp(ctx context.Context, mt *mtest.T, coll *mongo.Collection) {
+	mt.Helper()
+
+	result := coll.Database().RunCommand(ctx, bson.D{{"notARealCommand", true}})
+	assert.NotNil(mt, result.Err(), "expected RunCommand error, got: %v", result.Err())
+}
+
+func clamExplicitTruncLimitLogsFail(mt *mtest.T) []logTruncCaseValidator {
+	mt.Helper()
+
+	return []logTruncCaseValidator{
+		nil,
+		newLogTruncCaseValidator(mt, "failure", func(cmd string) error {
+			if len(cmd) != 5+len(logger.TruncationSuffix) {
+				return fmt.Errorf("expected reply to be %d bytes, got %d",
+					5+len(logger.TruncationSuffix), len(cmd))
+			}
+
+			return nil
+		}),
+	}
+
+}
+
+func TestCommandLoggingAndMonitoringProse(t *testing.T) {
+	t.Parallel()
+
+	mt := mtest.New(t, mtest.NewOptions().CreateClient(false))
+	defer mt.Close()
+
+	for _, tcase := range []struct {
+		// name is the name of the test case
+		name string
+
+		// collectionName is the name to assign the collection for
+		// processing the operations. This should be unique across test
+		// cases.
+		collectionName string
+
+		// maxDocumentLength is the maximum document length for a
+		// command message.
+		maxDocumentLength uint
+
+		// orderedLogValidators is a slice of log validators that should
+		// be 1-1 with the actual logs that are propagated by the
+		// LogSink. The order here matters, the first log will be
+		// validated by the 0th validator, the second log will be
+		// validated by the 1st validator, etc.
+		orderedLogValidators []logTruncCaseValidator
+
+		// operation is the operation to perform on the collection that
+		// will result in log propagation. The logs created by
+		// "operation" will be validated against the
+		// "orderedLogValidators."
+		operation func(context.Context, *mtest.T, *mongo.Collection)
+	}{
+		{
+			name:                 "1 Default truncation limit",
+			collectionName:       "46a624c57c72463d90f88a733e7b28b4",
+			operation:            clamDefaultTruncLimitOp,
+			orderedLogValidators: clamDefaultTruncLimitLogs(mt),
+		},
+		{
+			name:                 "2 Explicitly configured truncation limit",
+			collectionName:       "540baa64dc854ca2a639627e2f0918df",
+			maxDocumentLength:    5,
+			operation:            clamExplicitTruncLimitOp,
+			orderedLogValidators: clamExplicitTruncLimitLogs(mt),
+		},
+		{
+			name:                 "2 Explicitly configured truncation limit for failures",
+			collectionName:       "aff43dfcaa1a4014b58aaa9606f5bd44",
+			maxDocumentLength:    5,
+			operation:            clamExplicitTruncLimitFailOp,
+			orderedLogValidators: clamExplicitTruncLimitLogsFail(mt),
```

Comment:
I think it's still worth having a test for multi-byte character truncation if possible.

Reply:
I'm not really sure how to implement this, operation response sizes in bytes are very non-deterministic. There is virtual no guarantee the truncation would ever split in the middle of a code point. Perhaps we could create a massive string insert of multi-byte characters and then set the maxDocumentLength to like 10_000 or something and then make sure 1/50 tries truncates a MBC by excluding it?

To me this test seems redundant if we just add a test to the truncation routine that ensures this. The other tests already let us know that a truncations will occur.

Reply:
Done!